File: tests/python/test_workshop_schema.py
#!/usr/bin/env python3
"""契约测试：workshop_sites.json schema 校验。"""
import json
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent.parent
JSON_FILE = ROOT / "workshop_sites.json"
VALID_GROUPS = {"search", "github", "repo"}
REQUIRED_FIELDS = ["id", "icon", "label", "url", "desc", "group"]


def validate():
    errors = []

    if not JSON_FILE.exists():
        errors.append(f"MISSING: {JSON_FILE}")
        return errors, 0

    data = json.loads(JSON_FILE.read_text("utf-8"))
    if not isinstance(data, list) or len(data) == 0:
        errors.append("SCHEMA: must be a non-empty array")
        return errors, 0

    ids = set()
    for i, site in enumerate(data):
        prefix = f"[{i}] {site.get('id', '?')}"

        for field in REQUIRED_FIELDS:
            if field not in site:
                errors.append(f"{prefix}: missing required field '{field}'")
            elif not isinstance(site.get(field), str) or not site[field]:
                errors.append(f"{prefix}: '{field}' must be non-empty string")

        tid = site.get("id", "")
        if tid:
            if tid in ids:
                errors.append(f"{prefix}: duplicate id '{tid}'")
            ids.add(tid)

        group = site.get("group", "")
        if group and group not in VALID_GROUPS:
            errors.append(f"{prefix}: 'group' must be one of {VALID_GROUPS} (got '{group}')")

        # searchUrl if present must contain {{q}} or be a valid URL
        su = site.get("searchUrl", "")
        if su and "{{q}}" not in su and not su.startswith("http"):
            errors.append(f"{prefix}: 'searchUrl' should contain {{{{q}}}} or be a URL")

        # presetSearches if present
        ps = site.get("presetSearches", [])
        if ps:
            if not isinstance(ps, list):
                errors.append(f"{prefix}: 'presetSearches' must be an array")
            else:
                for j, p in enumerate(ps):
                    if not isinstance(p, dict) or "label" not in p:
                        errors.append(f"{prefix}: presetSearches[{j}] missing 'label'")

    return errors, len(data)


def main():
    errors, count = validate()
    if errors:
        print(f"FAILED: {len(errors)} schema violation(s) in {JSON_FILE.name}\n")
        for e in errors:
            print(f"  {e}")
        sys.exit(1)
    else:
        print(f"OK: {count} sites, all schema checks passed")


if __name__ == "__main__":
    main()

File: tests/test_workshop_schema.mjs
#!/usr/bin/env node
/**
 * 契约测试：workshop_sites.json schema 校验。
 * 由 tests/python/test_workshop_schema.py 迁移（2026-08-03），校验逻辑逐点保真。
 */
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');
const JSON_FILE = path.join(ROOT, 'workshop_sites.json');

const VALID_GROUPS = new Set(['search', 'github', 'repo']);
const REQUIRED_FIELDS = ['id', 'icon', 'label', 'url', 'desc', 'group'];

function validate() {
  const errors = [];

  if (!fs.existsSync(JSON_FILE)) {
    errors.push(`MISSING: ${JSON_FILE}`);
    return { errors, count: 0 };
  }

  let data;
  try {
    data = JSON.parse(fs.readFileSync(JSON_FILE, 'utf-8'));
  } catch (e) {
    errors.push(`SYNTAX: workshop_sites.json 解析失败: ${e.message}`);
    return { errors, count: 0 };
  }

  if (!Array.isArray(data) || data.length === 0) {
    errors.push('SCHEMA: must be a non-empty array');
    return { errors, count: 0 };
  }

  const ids = new Set();
  for (let i = 0; i < data.length; i++) {
    const site = data[i];
    const prefix = `[${i}] ${site?.id ?? '?'}`;

    for (const field of REQUIRED_FIELDS) {
      if (!(field in site)) {
        errors.push(`${prefix}: missing required field '${field}'`);
      } else if (typeof site[field] !== 'string' || !site[field]) {
        errors.push(`${prefix}: '${field}' must be non-empty string`);
      }
    }

    const tid = site?.id ?? '';
    if (tid) {
      if (ids.has(tid)) {
        errors.push(`${prefix}: duplicate id '${tid}'`);
      }
      ids.add(tid);
    }

    const group = site?.group ?? '';
    if (group && !VALID_GROUPS.has(group)) {
      errors.push(`${prefix}: 'group' must be one of ${[...VALID_GROUPS]} (got '${group}')`);
    }

    // searchUrl if present must contain {{q}} or be a valid URL
    const su = site?.searchUrl ?? '';
    if (su && !su.includes('{{q}}') && !su.startsWith('http')) {
      errors.push(`${prefix}: 'searchUrl' should contain {{q}} or be a URL`);
    }

    // presetSearches if present
    const ps = site?.presetSearches ?? [];
    if (ps && ps.length > 0) {
      if (!Array.isArray(ps)) {
        errors.push(`${prefix}: 'presetSearches' must be an array`);
      } else {
        for (let j = 0; j < ps.length; j++) {
          const p = ps[j];
          if (typeof p !== 'object' || p === null || !('label' in p)) {
            errors.push(`${prefix}: presetSearches[${j}] missing 'label'`);
          }
        }
      }
    }
  }

  return { errors, count: data.length };
}

const { errors, count } = validate();
if (errors.length) {
  console.error(`FAILED: ${errors.length} schema violation(s) in ${path.basename(JSON_FILE)}\n`);
  for (const e of errors) console.error(`  ${e}`);
  process.exit(1);
} else {
  console.log(`OK: ${count} sites, all schema checks passed`);
}
